Student/Student/Assignments/AssignmentList/AssignmentListViewController.swift

```diff
@@ -19,49 +19,287 @@
-        cell.textLabel?.text = presenter?.assignments[indexPath]?.name
-        cell.imageView?.image = .icon(.assignment, .line)
-        cell.imageView?.tintColor = color
+    func selectFirstCellOnIpad() {
+        if let splitView = splitViewController, splitView.viewControllers.count == 2 {
+            let ip = IndexPath(row: 0, section: 0)
+            tableView.selectRow(at: ip, animated: true, scrollPosition: .none)
+            tableView(tableView, didSelectRowAt: ip)
+        }
+    }
+
+    @IBAction func actionFilterClicked(_ sender: UIButton) {
+        if selectedGradingPeriod != nil {
+            filterByGradingPeriod(nil)
+        } else {
+            let alert = UIAlertController(title: nil, message: NSLocalizedString("Filter by:", comment: ""), preferredStyle: .actionSheet)
+
+            for period in gradingPeriods {
+                let action = AlertAction(period.title, style: .default) { [weak self] _ in
+                    self?.filterByGradingPeriod(period)
+                }
+                alert.addAction(action)
+            }
+            let cancel = UIAlertAction(title: NSLocalizedString("Cancel", comment: ""), style: .destructive, handler: nil)
+            alert.addAction(cancel)
+
+            alert.popoverPresentationController?.sourceRect = sender.bounds
+            alert.popoverPresentationController?.sourceView = sender
+
+            env.router.show(alert, from: self, options: .modal)
+        }
+    }
+
+    @objc func refresh(_ control: UIRefreshControl) {
+        fetchData()
+    }
+
+    func showSpinner(show: Bool = true) {
+        if show {
+            tableRefresher.beginRefreshing()
+        } else {
+            tableRefresher.endRefreshing()
+        }
+    }
+}
+
+extension AssignmentListViewController: UITableViewDataSource, UITableViewDelegate {
+    func assignment(for indexPath: IndexPath) -> APIAssignmentListAssignment? {
+        return assignments[ groups[indexPath.section].id.value ]?[indexPath.row]
+    }
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return groups.count
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        let loadCell = sectionHasNext[section] ? 1 : 0
+        let assignmentCnt = assignments[ groups[section].id.value ]?.count ?? 0
+        return assignmentCnt + loadCell
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell: ListCell = tableView.dequeue(for: indexPath)
+        let loadCell = sectionHasNext[indexPath.section]
+        if loadCell && indexPath.row == assignments[ groups[indexPath.section].id.value ]?.count ?? 0 {
+            cell.textLabel?.text = NSLocalizedString("Loading...", comment: "")
+            cell.detailTextLabel?.text = nil
+            cell.imageView?.image = nil
+
+            DispatchQueue.global().async { [weak self] in
+                self?.fetchData()
```

Comment:
This will get called every time `tableView_:cellForRowAt:)` gets called for this loading cell which could be several. We only want to load the first time it's called, right? I think you're going to need to track if you're currently loading for this `sectionHasNext[indexPath.section]` so it doesn't get called more times than necessary.